ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs

```diff
@@ -627,28 +636,202 @@ prop_generator_NonFailingBeararInfoScript (NonFailingBearerInfoScript s) = not (
+            return res
+                                                       )
+      `catch` (\(err :: SomeException) ->
+        -- Should error with NotReleasedListeningSockets
+        case (fromException err
+               :: Maybe (ResourceException (TestAddress Int))) of
+          Just _ ->
+            return (Right ())
+          Nothing ->
+            return (Left err)
+      )
+
+prop_connect_to_not_accepting_socket :: BearerInfoScript -> Property
+prop_connect_to_not_accepting_socket (BearerInfoScript script) =
+  prop_connect_to_template sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) $ \snocket ->
+        withAsync (server (TestAddress (0 :: Int)) snocket close loop)
+          $ \_ -> do
+            res <- client (TestAddress 1) (TestAddress 0) snocket close
+            case res of
+              -- Should timeout
+              Left _ -> return (Right ())
+              Right _ -> return (Left (toException UnexpectedOutcome))
+
+prop_connect_to_uninitialised_socket :: BearerInfoScript -> Property
+prop_connect_to_uninitialised_socket (BearerInfoScript script) =
+  prop_connect_to_template sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) $ \snocket -> do
+        res <- client (TestAddress (1 :: Int)) (TestAddress 0) snocket close
+        case res of
+          -- Should complain about no such listening socket
+          Left _ -> return (Right ())
+          Right _ -> return (Left (toException UnexpectedOutcome))
+
+prop_simultaneous_open :: BearerInfoScript -> Property
+prop_simultaneous_open (BearerInfoScript script) =
+  prop_connect_to_template sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) $ \snocket -> do
+        withAsync
+          (connectingServer (TestAddress (0 :: Int)) (TestAddress 1) snocket)
+          $ \clientAsync -> do
+            _ <- connectingServer (TestAddress 1) (TestAddress 0) snocket
+            wait clientAsync
 
 --
 -- Utils
 --
 
+server
+  :: ( MonadThread m
+    , MonadThrow m
+    )
+  => TestAddress addr
+  -> Snocket m fd (TestAddress addr)
+  -> (Snocket m fd (TestAddress addr) -> fd -> m ())
+  -> (m (Accept m fd (TestAddress addr)) -> m (b, Maybe fd))
+  -> m b
+server localAddress snocket closeF acceptF = do
+    labelThisThread "server"
+    bracket (open snocket TestFamily)
+            (closeF snocket)
+            (\fd -> do
+              bind snocket fd localAddress
+              listen snocket fd
+              (b, mbFd) <- acceptF (accept snocket fd)
+              traverse_ (close snocket) mbFd
+              return b
+            )
+
+loop :: MonadDelay m => a -> m b
+loop a = do
+  threadDelay 10000
+  loop a
+
+accept' :: MonadMask m => m (Accept m fd addr) -> m (Bool, Maybe fd)
```

Comment:
Let's rename it as `acceptOne`.

The boolean value is redundant, it would be better to return `Either fd IOError`, but `Maybe fd` is also fine if we don't need the error (e.g. for a counter example).